### PR TITLE
Fix algorithms for Persian and Hebrew calendars.

### PR DESCRIPTION
**Pull request by jon48**
Set Arabic to use civil epoch rather than astronomical one.
Add reciprocity jd <-> ymd tests.
Add average time performance indicator.

### CODE REVIEW

**Comment by jon48** [comment]
In the current code, several julian days produce invalid months:

- In the Jewish calendar, JD 381577 produces month 14 for instance.
       In the Jewish calendar, it seems those errors come from an estimation of jdToY not overestimated enough. Fix uses another estimation inspired from the Fourmilab converter.

- In the Jalali/Persian calendar, JD  2456373 produces month 13 for instance.
      I think the initial code was transposed from a javascript script, but there is a different behavior of the mod operator for negative numbers. Additionnally, for negative numbers, `(int) x` is different from `floor(x)`.

**Comment by jon48** [comment]
Concerning the Arabic calendar, the current algorithm is not wrong, but use the astronomical epoch (Thursday, 15 July 622 CE) instead of the civil one (Friday, 16 July 622 CE), which is more widely used in standard converter.
The existing code, was not entirely clear about it by the way, as it was setting `jdStart` to 1948440 which is the civil epoch, and some tests were assuming the civil values, causing trouble when trying to reciprocate.

More details here: https://www.staff.science.uu.nl/~gent0113/islam/islam_tabcal_others.htm

I can revert that part if necessary, but I think this is more inline with generally used converters.

File: test/ShimTest.php [hunk 1]
```diff
@@ -30,6 +30,12 @@ class ShimTest extends TestCase {
 	 */
 	const LARGE_PRIME = 235741;
 
+	
+	protected function setUp()
```
Comment (fisharebest) [inline test/ShimTest.php]: Do we need this empty function?

File: src/Shim.php [hunk 2]
```diff
@@ -751,4 +751,28 @@ public static function unixToJd($timestamp) {
...
+	 *    (Shim)           Shim::mod(-3, 4) = 1
+	 * 
+	 * @param number $dividend
+	 * @param number $divisor
+	 * @return number
+	 */
+	public static function mod($dividend, $divisor) {
+	    if ($dividend == 0 || $divisor == 0) return 0;
+	    	    
+	    // negative case
```
Comment (fisharebest) [inline src/Shim.php]: Is this very inefficient for large numbers?  e.g. `mod(-999999999,10)`?

Perhaps something like this?

```
$modulus = $dividend % $divisor;
if ($modulus < 0) {
    $modulus += $divisor;
}

return $modulus;
}

File: src/Shim.php [hunk 3]
```diff
@@ -751,4 +751,28 @@ public static function unixToJd($timestamp) {
...
+	 * For example: 
+	 *    (Standard PHP)   -3 % 4 = -3
+	 *    (Shim)           Shim::mod(-3, 4) = 1
+	 * 
+	 * @param number $dividend
+	 * @param number $divisor
+	 * @return number
+	 */
+	public static function mod($dividend, $divisor) {
+	    if ($dividend == 0 || $divisor == 0) return 0;
```
Comment (fisharebest) [inline src/Shim.php]: `$dividend == 0` is not a problem.  For efficiency, remove this check?

Can we use `===` instead of `==`?

**Comment by fisharebest** [comment]
Hi Jon - thanks for this!

I have been looking for another implementation of the arabic algorithm to use as a reference.

I've added a couple of minor comments.  Can you also reformat using the existing code style?

* tabs for indent, not spaces
* no trailing whitespace
* always use `{` and `}` for `if/while`
* space between `if/while` and `(`
* blank line between `return` and previous statement.

File: test/ShimTest.php (hunk 1 above)
Comment (jon48) [inline test/ShimTest.php]: No, did not realise it had been introduced... I removed it

File: src/Shim.php (hunk 2 above)
Comment (jon48) [inline src/Shim.php]: Indeed. I did not hit the issue, as the algorithms do not produce large numbers, but that is way more clever. Done in my latest commit.

**Comment by fisharebest** [comment]
I've merged this PR.  Thanks for your help!

I made a few changes:
You used the latest PHPUnit - which does not work with old PHP.
The "reciprocity" tests took too long on TravisCI, so I reduced them.
I moved the `mod()` function to the class where it is used.

**Pull request closed without merging**